### PR TITLE
OP-488 Implement patient search in oh-core

### PR DESCRIPTION
**Pull request by npasquetto**
The aim of this implementation is to add **basic search capabilities** on patient entity to support OH 20 UI requirements (https://openhospital.atlassian.net/browse/OP-323).

At the moment, supported search parameters are:

- birthdate
- address
- firstName
- secondName

This is [the first of two PR](https://openhospital.atlassian.net/browse/OP-487): the second one will be [on the oh-api project](https://openhospital.atlassian.net/browse/OP-489), but it _depends_ on this one.

### CODE REVIEW

File: src/main/java/org/isf/patient/manager/PatientBrowserManager.java [hunk 1]
```diff
@@ -21,11 +21,6 @@
  */
 package org.isf.patient.manager;
 
-import java.util.ArrayList;
```
Comment (dbmalkovsky) [inline src/main/java/org/isf/patient/manager/PatientBrowserManager.java:24]: Please see the **Code style** section of the repo README file to enable the project's code style rules.  For example, package imports are not supported and import ordering is

- statics
- java
- javax
- org
- com

Thanks.

File: src/test/java/org/isf/patient/test/Tests.java [hunk 2]
```diff
@@ -216,6 +223,17 @@ public void testIoGetPatientAll() throws Exception {
...
 
+	@Test
+	public void testIoGetPatientsByParams() throws Exception {
+		_setupTestPatient(false);
+		Map<String, Object> params = new HashMap<String, Object>();
+		params.put("firstName", "TestFirstN");
+		params.put("birthDate", new GregorianCalendar(1984, Calendar.AUGUST, 14).getTime());
+		params.put("address", "TestAddress");
+		ArrayList<Patient> patients = patientIoOperation.getPatients(params);
+		assertThat(patients.size() > 0).isTrue();
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/patient/test/Tests.java:234]: A more idiomatic version would be:  **assertThat(patients.size()).isPositive();**

File: src/test/java/org/isf/patient/test/Tests.java [hunk 3]
```diff
@@ -48,6 +48,13 @@
 import org.springframework.data.domain.PageRequest;
 import org.springframework.data.domain.Pageable;
 
+import java.io.File;
+import java.lang.reflect.Field;
+import java.util.*;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.assertj.core.api.Assertions.assertThatThrownBy;
```
Comment (dbmalkovsky) [inline src/test/java/org/isf/patient/test/Tests.java:56]: I don't believe all the imports added are actually used in the file, are they?  For example, where is **assertThatThrownBy** or **java.io.File**?

File: src/main/java/org/isf/patient/manager/PatientBrowserManager.java (hunk 1 above)
Comment (npasquetto) [reply src/main/java/org/isf/patient/manager/PatientBrowserManager.java]: Sorry, I thought I had the correct settings.
Now it should be ok.

File: src/test/java/org/isf/patient/test/Tests.java (hunk 2 above)
Comment (npasquetto) [reply src/test/java/org/isf/patient/test/Tests.java]: Yes! I've modified it.

File: src/test/java/org/isf/patient/test/Tests.java (hunk 3 above)
Comment (npasquetto) [reply src/test/java/org/isf/patient/test/Tests.java]: It seems to me that they are used.

File: src/main/java/org/isf/patient/service/PatientIoOperationRepositoryImpl.java [hunk 4]
```diff
@@ -90,15 +91,48 @@
...
+		for (Map.Entry<String, Object> entry : params.entrySet()) {
+			Path<String> keyPath = patient.get(entry.getKey());
+
+			if (entry.getKey().equals("birthDate")) {
+				Date birthDateFrom = (Date) entry.getValue();
+				Date birthDateTo = new Date(birthDateFrom.getTime() + (1000 * 60 * 60 * 24));
+				predicates.add(cb.between(keyPath.as(Date.class), birthDateFrom, birthDateTo));
+			} else {
+				if (entry.getValue() instanceof String) {
+					predicates.add(cb.like(keyPath, (String) entry.getValue() + "%"));
```
Comment (mwithi) [inline src/main/java/org/isf/patient/service/PatientIoOperationRepositoryImpl.java:128]: I think we should implement case insensitive `like` as done in `wordExistsInOneOfPatientFields` method, eventually reusing the same code

**Review by mwithi — CHANGES_REQUESTED** [review]
To implement case insensitive `like`

**Review by mwithi — APPROVED** [review]
In the end:

`getPatientsByOneOfFieldsLike(String keyword)` searches a keyword in any field of a "hard-coded" set of fields

`getPatientsByParams(Map<String, Object> params)` searches a specific set of fields-values in AND relation each other

I think both methods make sense. Thanks a lot!

**Pull request merged**